Fix TabReader error handling

lkuchenb opened the pull request:
There are no guarantees about `errno` when there is _no_ error. I ran into a situation where `errno==ENOMEM` on our system (for unknown reasons, unfortunately) which definately didn't come from `strtol()`.

It triggered the "your scan number is not an integer" error and percolator quit. This patch fixes the error handling for `strtol()` and we're not experiencing any problems with it anymore, however, I remain puzzled how `errno` was set to `ENOMEM` 🤷‍♂️

MatthewThe commented:
Thanks for the pull request! This is indeed a corner case that we are currently not handling well.

We need to add `errno=0` to the other `reader.readX()` functions as well. Furthermore, I think that the reason why we did not do this previously, was to make sure that even when a subsequent `strtox()` function worked fine, we would still raise an error the next time we check `reader.error()`. In this way, we can do multiple `reader.readX()` calls, but only have to check `reader.error()` once at the end. 

To preserve this behavior we should only set `err = errno` if `errno != 0`. Unfortunately, this might sometimes give an error message that is not actually applicable to the actual problem (as you have experienced), but it will at least show us the approximate part of the code that causes the error.

lkuchenb commented:
> We need to add errno=0 to the other reader.readX() functions as well.

Indeed! I added it for `strtod()` and removed `errno` handling from the `strchr()` call since it doesn't even set `errno`.

> Furthermore, I think that the reason why we did not do this previously, was to make sure that even when a subsequent strtox() function worked fine, we would still raise an error the next time we check reader.error(). In this way, we can do multiple reader.readX() calls, but only have to check reader.error() once at the end. To preserve this behavior we should only set `err = errno` if `errno != 0`

I added the conditional assignment, too.

MatthewThe commented:
Thanks a lot! I will test this out after I finish with the other pull request.
